lib/farscape/plugins.rb

```diff
@@ -2,59 +2,92 @@
-      (disabled_plugins & [options[:name],options[:type]]).first
+      options = normalize_selector(options)
+      return @plugins[options[:name]][:enabled] if options.include?([:name])
+      why_disabled(options).empty? ? false : true
+    end
+    
+    def enabled?(options)
+      !disabled?(options)
     end
 
-    # Prevents a plugin [type] from being registered, and unregisters it if it's already there
-    def disable(name_or_type)
-      disabled_plugins << name_or_type
-      plugins_to_disable, @plugins = @plugins.partition { |plugin| [plugin[:name], plugin[:type]].include?(name_or_type) }
-      plugins_to_disable.each { |plugin| disable_plugin(plugin) }
+    # Prevents a plugin from being registered, and disables it if it's already there
+    def disable!(name_or_type)
+      name_or_type = normalize_selector(name_or_type)
+      @disabling_rules << name_or_type
+      selected_plugins = find_attr_intersect(@plugins, name_or_type)
+      selected_plugins.each { |plugin| @plugins[plugin][:enabled] = false }
     end
 
+    # Allows a plugin to be registered, and enables it if it's already there
+    def enable!(name_or_type)
+      name_or_type = normalize_selector(name_or_type)
+      @disabling_rules.delete(name_or_type)
+      selected_plugins = find_attr_intersect(@plugins, name_or_type)
+      selected_plugins.each { |plugin| @plugins[plugin][:enabled] = true }
+    end
+
+    # Returns the Poset representing middleware dependency
+    def middleware_stack
+      stack = PartiallyOrderedList.new { |m,n| order_middleware(m,n) }
+      enabled_plugins.map do |_, plugin|
+        [*plugin[:middleware]].map do |middleware|
+          middleware = {class: middleware} unless middleware.is_a?(Hash)
+          middleware[:type] = plugin[:type]
+          middleware[:plugin] = plugin[:name]
+          stack.add(middleware)
+        end
+      end
+      stack
+    end
+    
     # Removes all plugins and disablings of plugins
     def clear
-      plugins.each { |pi| disable_plugin(pi) }
-      @plugins = []
-      @disabled_plugins = []
+      @plugins = {}
+      @disabling_rules = []
     end
 
     private
 
-    # Handles the state cleanup when an existing plugin is removed.
-    # Currently all this has to do is delete middleware but that'll change soon.
-    def disable_plugin(plugin)
-      @middleware_stack.select{ |m| m[:plugin] == plugin[:name] }.each { |m| @middleware_stack.delete(m) }
-    end
-
-    # Adds a hash representing each middleware class and where it came from to the middleware stack
-    def add_middleware(options)
-      [*options[:middleware]].each do |middleware|
-        middleware = {class: middleware} unless middleware.is_a?(Hash)
-        middleware[:type] = options[:type]
-        middleware[:plugin] = options[:name]
-        @middleware_stack.add(middleware)
-      end
+    def normalize_selector(name_or_type)
+      name_or_type.is_a?(Hash) ? name_or_type : { name: name_or_type, type: name_or_type}
```

Comment:
I'd like our documentation (and actual plugins) to consistently show names as being CamelCase and types as being snake_case to prevent collision.